File: desktop/src/Qlipq.Ffmpeg/Progress.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace Qlipq.Ffmpeg;

public sealed record ProgressUpdate
{
    /// <summary>Output timestamp reached so far, in seconds, or null if unknown.</summary>
    public double? OutTimeSec { get; set; }

    /// <summary>True once ffmpeg reports <c>progress=end</c>.</summary>
    public bool Done { get; set; }
}

public static partial class Progress
{
    [GeneratedRegex(@"^(\d+):(\d{2}):(\d{2}(?:\.\d+)?)$", RegexOptions.ECMAScript)]
    private static partial Regex TimecodeRegex();

    [GeneratedRegex(@"\r?\n")]
    private static partial Regex LineSplitRegex();

    /// <summary>Parse a <c>HH:MM:SS.micro</c> timecode into seconds, or null if unparseable.</summary>
    public static double? ParseTimecode(string value)
    {
        var match = TimecodeRegex().Match(value.Trim());
        if (!match.Success) return null;
        var hours = double.Parse(match.Groups[1].Value, CultureInfo.InvariantCulture);
        var minutes = double.Parse(match.Groups[2].Value, CultureInfo.InvariantCulture);
        var seconds = double.Parse(match.Groups[3].Value, CultureInfo.InvariantCulture);
        return hours * 3600 + minutes * 60 + seconds;
    }

    /// <summary>
    /// Parse one or more <c>-progress pipe:1</c> chunks. ffmpeg emits <c>key=value</c> lines in
    /// blocks terminated by <c>progress=continue|end</c>; we return the latest timestamp seen and
    /// whether the run finished. Note: ffmpeg's <c>out_time_ms</c> is actually microseconds — both
    /// it and <c>out_time_us</c> are divided by 1,000,000 (do not "fix" this).
    /// </summary>
    public static ProgressUpdate ParseProgress(string text)
    {
        double? outTimeSec = null;
        var done = false;

        foreach (var raw in LineSplitRegex().Split(text))
        {
            var line = raw.Trim();
            var eq = line.IndexOf('=');
            if (eq < 0) continue;
            var key = line[..eq];
            var value = line[(eq + 1)..];

            if (key is "out_time_us" or "out_time_ms")
            {
                if (double.TryParse(value.Trim(), NumberStyles.Float, CultureInfo.InvariantCulture, out var micros)
                    && double.IsFinite(micros) && micros >= 0)
                {
                    outTimeSec = micros / 1_000_000;
                }
            }
            else if (key == "out_time")
            {
                var parsed = ParseTimecode(value);
                if (parsed is not null) outTimeSec = parsed;
            }
            else if (key == "progress")
            {
                done = value == "end";
            }
        }

        return new ProgressUpdate { OutTimeSec = outTimeSec, Done = done };
    }

    /// <summary>Clamp an export progress fraction (0..1) from the current and total seconds.</summary>
    public static double ProgressFraction(double? outTimeSec, double durationSec)
    {
        if (outTimeSec is null || durationSec <= 0) return 0;
        return Math.Min(1, Math.Max(0, outTimeSec.Value / durationSec));
    }
}

File: iced/crates/qlipq-ffmpeg/src/progress.rs
use std::sync::LazyLock;

use regex::Regex;

#[derive(Debug, Clone, PartialEq)]
pub struct ProgressUpdate {
    /// Output timestamp reached so far, in seconds, or `None` if unknown.
    pub out_time_sec: Option<f64>,
    /// True once ffmpeg reports `progress=end`.
    pub done: bool,
}

static TIMECODE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^([0-9]+):([0-9]{2}):([0-9]{2}(?:\.[0-9]+)?)$").unwrap());

/// Parse a `HH:MM:SS.micro` timecode into seconds, or `None` if unparseable.
pub fn parse_timecode(value: &str) -> Option<f64> {
    let caps = TIMECODE.captures(value.trim())?;
    let h: f64 = caps[1].parse().ok()?;
    let m: f64 = caps[2].parse().ok()?;
    let s: f64 = caps[3].parse().ok()?;
    Some(h * 3600.0 + m * 60.0 + s)
}

/// Parse one or more `-progress pipe:1` chunks. ffmpeg's `out_time_ms` is actually microseconds —
/// both it and `out_time_us` are divided by 1,000,000 (do not "fix" this).
pub fn parse_progress(text: &str) -> ProgressUpdate {
    let mut out_time_sec: Option<f64> = None;
    let mut done = false;

    for raw in text.split('\n') {
        let line = raw.trim();
        let Some(eq) = line.find('=') else { continue };
        let key = &line[..eq];
        let value = &line[eq + 1..];

        match key {
            "out_time_us" | "out_time_ms" => {
                if let Ok(micros) = value.trim().parse::<f64>() {
                    if micros.is_finite() && micros >= 0.0 {
                        out_time_sec = Some(micros / 1_000_000.0);
                    }
                }
            }
            "out_time" => {
                if let Some(parsed) = parse_timecode(value) {
                    out_time_sec = Some(parsed);
                }
            }
            "progress" => {
                done = value == "end";
            }
            _ => {}
        }
    }

    ProgressUpdate { out_time_sec, done }
}

/// Clamp an export progress fraction (0..1) from the current and total seconds.
pub fn progress_fraction(out_time_sec: Option<f64>, duration_sec: f64) -> f64 {
    match out_time_sec {
        Some(t) if duration_sec > 0.0 => (t / duration_sec).clamp(0.0, 1.0),
        _ => 0.0,
    }
}
